<a href="https://colab.research.google.com/github/Dkepffl/2022-2-ESAA/blob/main/Assignment/Assignment221007_text1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CHAPTER 08 텍스트 분석**
___
- 머신러닝이 보편화되면서 NLP와 텍스트 분석을 구분하는 것이 큰 없어 보이지만, 굳이 구분하자면 
  + NLP는 머신이 인간의 언어를 이해하고 해석하는 데 더 중점을 둔다.
  + 텍스트 마이닝이라고도 불리는 텍스트 분석은 비정형 텍스트에서 의미 있는 정보를 추출하는 것에 좀 더 중점을 둔다.
- NLP 기술이 발전함에 따라 텍스트 분석도 정교하게 발전할 수 있다. 
- NLP는 언어를 해석하기 위한 **기계 번역**, 자동으로 질문을 해석하고 답을 해주는 **질의응답 시스템** 등을 수행한다.
- 텍스트 분석은 머신러닝, 언어 이해, 통계 등을 활용해 모델을 수립하고 정보를 추출해 **비즈니스 인텔리전스**나 **예측 분석** 등의 분석 작업을 주로 수행한다.
  + **텍스트 분류(Text Classification)** : 문서가 특정 분류 또는 카테고리에 속하는 것을 예측하는 기법을 통칭한다. 스팸 메일 검출이나 기사 카테고리 분류 등을 예로 들 수 있으며, 지도학습을 적용한다.
  + **감성 분석(Sentiment Analysis)** : 텍스트에서 나타나는 감정/판단/믿음/의견/기분 등의 주관적인 요소를 분석하는 기법을 총칭한다. 지도학습뿐만 아니라 비지도 학습을 적용할 수도 있다.
  + **텍스트 요약(Summarization)** : 텍스트 내에서 중요한 주제나 중심 사상을 추출하는 기법. 대표적으로 토픽 모델링(Topic Modeling)이 있다.
  + **텍스트 군집화(Clustering)오ㅘ 유사도 측정** : 비슷한 유형의 문서에 대해 군집화를 수행하는 기법. 텍스트 분류를 비지도 학습으로 수행하는 방법 중 하나이다. 유사도 측정 역시 문서들 간의 유사도를 측정해 비슷한 문서끼리 모을 수 있는 방법이다.

## **01 텍스트 분석 이해**
___
- 텍스트 분석은 비정형 데이터인 텍스트를 분석하는 것이다.
- 지금까지 ML 모델은 주어진 정형 데이터를 기반으로 모델을 수립하고 예측했다. 머신러닝 알고리즘은 수자형의 피처 기반 데이터만 입력받을 수 있기 때문에, 텍스트를 머신러닝에 적용하기 위해서는 **'비정형 텍스트 데이터를 어떻게 피처 형태로 추출하고 추출된 피처에 의미 있는 값을 부여하는가'**가 중요한 문제입니다.
- **피처 벡터화(혹은 피처 추출)** : 텍스트를 word 기반의 다수의 피처로 추출하고, 이 피처에 단어 빈도수와 같은 숫자값을 부여하여, 텍스트를 단어의 조합인 벡터값으로 표현하는 것.
  + 대표적인 방법으로 BOW(Bag of Words)와 Word2Vec 방법이 있다.


### | **텍스트 분석 수행 프로세스**
- 텍스트 분석 수행 프로세스는 다음과 같다.
  1. **텍스트 사전 준비작업(텍스트 전처리)** : 텍스트를 피처로 만들기 전에 미리 클렌징, 대/소문자 변경, 특수 문자 삭제 등의 **클렌징 작업**, 단어(Word) 등의 **토큰화 작업**, **의미 없는 단어(Stop Word) 제거 작업**, 어근 추출(Stemming/Lemmatization) 등의 **텍스트 정규화 작업**
  2. **피처 벡터화/추출** : 사전 준비 작업으로 가공된 텍스트에서 피처를 추출하고, 여기에 벡터 값을 할당한다. 대표적인 방법으로, BOW와 Word2Vec이 있다.
  3. **ML 모델 수립 및 학습/예측/평가** : 피처 벡터화 된 데이터셋에서 ML 모델을 적용해 학습/예측 및 평가를 수행한다.


### | **파이썬 기반의 NLP, 텍스트 분석 패키지**
- **NLTK(Natural Language Toolkit for Python)** : 파이썬의 가장 대표적인 NLP 패키지. 방대한 데이터셋과 서브 모듈을 가지고 있으며, NLP의 거의 모든 영역을 커버하고 있다. 많은 NLP 패키지가 NLTK의 영향을 받아 작성되고 있다. 수행 속도가 느려 대규모의 데이터셋에서는 제대로 활용되지 못하고 있다.
- **Gensim** : 토픽 모델링 분야에서 가장 두각을 나타내는 패키지
- **SpaCy** : 뛰어난 수행 성능으로 최근 가장 주목을 받는 NLP 패키지
- 사이킷런과 위의 NLP 전용 패키지를 결합해 텍스트 분석을 진행한다.


## **02 텍스트 사전 준비 작업(텍스트 전처리)-텍스트 정규화**
___
- 텍스트 자체를 바로 피처로 만들 수는 없다. 사전에 텍스트를 가공하는 준비 작업이 필요하다.
- **텍스트 정규화**는 텍스트를 머신러닝 알고리즘이나 NLP 애플리케이션에 입력 데이터로 사용하기 위해 클렌징, 정제, 토큰화, 어근화 등의 다양한 텍스트 데이터의 사전 작업을 수행하는 것이다.

### | **클렌징**
- 텍스트에서 분석에 방해가 되는 불필요한 문자, 기호 등을 사전에 제거하는 작업니다. 예를 들어 HTML, XML 태그나 특정 기호 등을 제거한다.


### | **텍스트 토큰화**
- 문서에서 문장을 분리하는 문장 토큰화와 문장에서 다어를 토큰으로 분리하는 단어 토큰화로 나눌 수 있다. NLTK는 텍스트 토큰화를 위한 다양한 API를 제공한다.


#### **문장 토큰화**
- 문장 토큰화는 문장의 마침표(.), 개행문자(\n) 등 문장의 마지막을 뜻하는 기호에 따라 분리하는 것이 일반적이다. 혹은 정규 표현식에 따른 문장 토큰화도 가능하다.
- NTLK에서 일반적으로 많이 쓰이는 `sent_tokenize`를 이용해 토큰화를 수행해보자. 3개의 문장으로 이루어진 텍스트 문서를 문장으로 각각 분리할 것이다.

In [1]:
from nltk import sent_tokenize
import nltk
nltk.download('punkt') # 마침표, 개행 문자 등의 데이터셋을 다운로드

# 예시 문서
text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
               You can see it out your window or on your television. \
               You feel it when you go to work, or go to church or pay your taxes.'

# 토큰화 수행               
sentences = sent_tokenize(text=text_sample)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
print(type(sentences),len(sentences))

<class 'list'> 3


In [3]:
print(sentences)

['The Matrix is everywhere its all around us, here even in this room.', 'You can see it out your window or on your television.', 'You feel it when you go to work, or go to church or pay your taxes.']


- `sent_tokenize()`는 각각의 문장으로 구성된 list 객체를 반환한다.
- 리스트의 길이는 3으로 3개의 문장으로 된 문자열을 가지고 있는 리스트 객체가 반환되었다.

#### **단어 토큰화**
- 단어 토근화(Word Tokenization)는 단어로 토큰화하는 것이다. 기본적으로 공백, 콤마(,), 마침표(.), 개행 문자 등으로 단어를 분리하지만, 역시 정규 표현식을 이용해 다양한 유형으로 토큰화를 수행할 수 있다.
- 마침표(.)나 개행 문자와 같이 문장을 분리하는 구분자를 이용해 단어를 토큰화할 수 있으므로, Bag of Word와 같이 단어의 순서가 중요하지 않은 경우, 문장 토큰화를 사용하지 않고, 단어 토큰화만 사용한다.
- 일반적으로 문장 토큰화는 각 문작이 가지는 구조적 의미가 중요할 때 사용한다.
- NTLK에서 제공하는 `word_tokenzie()`를 이용해 단어로 토큰화해보자.

In [4]:
from nltk import word_tokenize

sentence = "The Matrix is everywhere its all around us, here even in this room."
words = word_tokenize(sentence)

In [5]:
print(type(words), len(words))

<class 'list'> 15


In [6]:
print(words)

['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.']


- 이번에는 `sent_tokenize`와 `word_tokenize`를 조합해 문서에 대해서 모든 단어를 토큰화해보자. 문장 토큰화 예제에서 사용했던 문장에 문장별로 단어 토큰화를 적용한다.
- 문서를 먼저 문장으로 나누고, 개별 문장을 다시 단어로 토큰화하는 `tokenize_text()` 함수를 생성해보자.

In [7]:
from nltk import word_tokenize, sent_tokenize

def tokenize_text(text):
    # 문장별로 분리
    sentences = sent_tokenize(text)

    # 개별 문장을 단어로 토큰화
    word_tokens = [word_tokenize(sentence) for sentence in sentences]
    return word_tokens

In [8]:
word_tokens = tokenize_text(text_sample) # 토근화 수행
print(type(word_tokens),len(word_tokens))

<class 'list'> 3


In [9]:
print(word_tokens)

[['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.'], ['You', 'can', 'see', 'it', 'out', 'your', 'window', 'or', 'on', 'your', 'television', '.'], ['You', 'feel', 'it', 'when', 'you', 'go', 'to', 'work', ',', 'or', 'go', 'to', 'church', 'or', 'pay', 'your', 'taxes', '.']]


- 문장을 단어별로 하나씩 토큰화할 겨우, 문맥적인 의미는 무시된다. 이러한 문제를 조금이라도 해결하기 위해 **n-graom**을 도입하였다.
  + n-gram은 연속된 n개의 단어를 하나의 토큰화 단위로 분리하는 것이다.
  - n개 단어 크기의 윈도우를 만들어 문장의 처음부터 오른쪽으로 움직이며 토큰화를 수행한다.

### | **스톱 워드 제거**
- **스톱 워드(Stop Word)**는 분석에 큰 의미가 없는 단어를 의미한다. 영어에서 be 동사나 관사와 같이 문장을 구성하는 필수 문법 요소지만, 문맥적으로 큰 의미가 없는 단어가 이에 해당한다.
- 이 단어의 경우, 문법적인 특성으로 인해 빈번하게 텍스트에 나타나 제거하지 않으면 중요한 단어로 인지될 수 있다. 따라서 이 의미 없는 단어를 제거하는 것은 중요한 전처리 작업이다.
- NLTK는 가장 다양한 언어의 스톱 워드를 제공한다. 먼저 NLTK의 `stopwords` 목록을 다운받아 보자.

In [10]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [11]:
# 영어의 stopwords 개수 확인
print('영어 stop words 개수:',len(nltk.corpus.stopwords.words('english')))
print(nltk.corpus.stopwords.words('english')[:20])

영어 stop words 개수: 179
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his']


- 영어의 경우, 스톱 워드의 개수가 179개이다.
- 앞에서 문장별로 단어를 토근화해 생성한 `word_tokens` 리스트에 stopwords를 필터링으로 제거해 의미 있는 단어만 추출해보자.

In [12]:
import nltk

In [13]:
stopwords = nltk.corpus.stopwords.words('english')
all_tokens = []

for sentence in word_tokens:
    filtered_words=[]
    for word in sentence:
        word = word.lower() # 소문자로 변환
        if word not in stopwords:
            filtered_words.append(word) # 주어진 word가 stopwords에 해당하지 않으면 리스트에 추가
    all_tokens.append(filtered_words)

In [14]:
print(all_tokens)

[['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['see', 'window', 'television', '.'], ['feel', 'go', 'work', ',', 'go', 'church', 'pay', 'taxes', '.']]


- is, this와 같은 스톱 워드가 제거되었다.

### | **Stemming과 Lemmatization**
- 많은 언어에서 문법적인 요소에 따라 단어가 다양하게 변화한다. **Stemming**과 **Lemmatization**은 이렇게 문법적 또는 의미적으로 변화하는 단어의 원형을 찾는 작업이다.
- 두 가지 모두 단어의 원형을 찾는다는 목적은 유사하지만, Lemmatization이 Stemming보다 정교하며, 의미론 기반에서 단어의 원형을 찾는다.
  + Stemming은 원형 단어로 변환 시 일반적인 방법을 적용하거나 더 단순화된 방법을 적용해, 원래 단어에서 일부 철자가 훼손된 어근 단어를 추출하는 경향이 있다.
  + Lemmatization은 품사와 같은 문법 요소와 의미적인 부분을 감안해 정환한 철자로 된 어근 단어를 찾아준다. 따라서 Lemmatization이 Stemming보다 변환에 더 오랜 시간을 피룡로 한다.
- NLTK는 Porter, Lancaster, Snowball Stemmer 등 다양한 Stemmer를 제공한다.
- Lemmatization을 위해서는 WordNetLemmatizer를 제공한다.



In [15]:
# Stemmer 예제-Lancaster
from nltk.stem import LancasterStemmer
stemmer = LancasterStemmer() # Stemmer 객체 생성

print(stemmer.stem('working'),stemmer.stem('works'),stemmer.stem('worked')) # stem 메서드 호출
print(stemmer.stem('amusing'),stemmer.stem('amuses'),stemmer.stem('amused'))
print(stemmer.stem('happier'),stemmer.stem('happiest'))
print(stemmer.stem('fancier'),stemmer.stem('fanciest'))

work work work
amus amus amus
happy happiest
fant fanciest


- 진행형, 3인칭 단수, 과거형에 따른 동사, 그리고 비교급, 최상급에 따른 형용사의 변화에 따라 Stemming은 조금 단순하게 원형 단어를 찾아준다.
- work의 경우, 진행형, 3인칭 단수, 과겨형 모두 원형 단어에 ing, s, ed가 붙는 단순한 변화이기 때문에, work를 제대로 인식한다.
- 하지만 ammuse의 경우 amus를 원형으로 인식하였으며, 형용사인 happy, fancy의 경우, 비교형, 최상급형으로 변형된 단어의 원형을 정확하게 찾지 못했다.


In [21]:
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [22]:
lemma = WordNetLemmatizer()
print(lemma.lemmatize('amusing','v'),lemma.lemmatize('amuses','v'),lemma.lemmatize('amused','v'))
print(lemma.lemmatize('happier','a'),lemma.lemmatize('happiest','a'))
print(lemma.lemmatize('fancier','a'),lemma.lemmatize('fanciest','a'))

amuse amuse amuse
happy happy
fancy fancy


- 앞의 Stemmer보다 정확하게 원형 단어를 추출하였다.

## **03 Bag of Words-BOW**
___
- Bag of Words 모델은 문서가  가지는 모든 단어(Words)를 문맥이나 순서를 무시하고 일괄적으로 단어에 대해 빈도 값을 부여하여, 피처 값을 추출하는 모델이다.
- 문서 내 모든 단어를 한꺼번에 봉투(Bag) 안에 넣고 흔들어서 섞는다는 의미로 Bag of Words(BOW) 모델이라고 한다.
- 다음과 같은 2개의 문장이 있다고 가정하고 BOW의 단어 수(Word Count) 기반으로 피처를 추출해보자.
```
문장 1 : 'My wife likes to watch baseball games and my daugther likes to watch baseball games too'
문장 2 : 'My wife likes to play baseball'
```
  1. 문장 1과 문장 2에 있는 모든 단어에서 중복을 제거하고, 각 단어를 컬럼 형태로 나열한다. 그리고 각 단어에 고유의 인덱스를 부여한다.
  2. 개별 문장에서 해당 단어가 나타나는 횟수(Occurrence)를 각 단어(단어 인덱스)에 기재한다.
- BOW 모델의 장점은 쉽고 빠르게 구축할 수 있다는 것이다. 단순히 단어의 발생 횟수에 기반하고 있지만, 예상보다 문서의 특징을 잘 나타낼 수 있는 모델이어서, 여러 분야에서 활용도가 높다.
- 하지만 BOW 기반의 NLP 연구는 다음과 같은 제약에 부딪히고 있다.
  + **문맥 의미(Semantic Context) 반영 부족** : 단어의 순서를 고래하지 않아, 문장 내에서 단어의 문맥적인 의미가 무시된다. 
  + **희소 행렬 문제(희소성, 희소 행렬)** : BOW로 피처 벡터화를 수행하면 희소 행렬 형태의 데이터셋이 만들어지기 쉽다. 희소 행렬은 일반적으로 ML 알고리즘의 수행 시간과 예측 성능을 떨어뜨린다. 이를 해결하기 위해 희소 행렬을 위한 특별한 기법이 존재한다.

### | **BOW 피처 벡터화**
- **피처 벡터화** : 텍스트를 특정 의미를 가지는 숫자형 값인 벡터 값으로 변환하는 것
- 각 문서의 텍스트를 단어로 추출해 피처로 할당하고, 각 단어의 발생 빈도수 등을 값으로 부여해 각 문서를 이 단어 피처의 발생 빈도 값으로 구성된 벡터로 만들 수 있다.
- BOW 모델에서는 모든 문서에서 모든 단어를 컬럼형태로 나열하고 각 문서에서 해당 단어의 횟수나 정규화된 빈도를 값으로 부여하는 데이터셋 모델로 변경하여 피처 벡터화를 수행한다.
- BOW의 피처 벡터화는 두 가지 방식이 있다.
  + **카운트 기반의 벡터화** : 단어 피처에 값을 부여할 때 각 문어에서 해당 단어가 나타나는 횟수(Count)를 부여한다.
  + **TF-IDF(Term Frequency-Inverse Document Frequency) 기반의 벡터화** : 개별 문서에서 자주 나타나는 단어에 높은 가중치를 주되, 모든 문서에서 전반적으로 자주 나타나는 단어에 대해서는 페널티를 준다.
- 문서마다 텍스트가 길고 문서의 개수가 많은 경우 카운트 방식보다는 TF-IDF 방식을 사용하는 것이 더 좋다.

### | **사이킷런의 Count 및 TF-IDF 벡터화 구현 : CountVectorizer, TfidVectorizer**
- 사이킷런의 `CountVectorizer` 클래스는 카운트 기반의 벡터화를 구현한 클래스이다.
  + 피처 벡터화뿐만 아니라, 소문자 일괄 변환, 토큰화, 스톱 워드 필터링 등 텍스트 전처리도 함께 수행한다.
  + `fit()`과 `transform()`을 통해 피처 벡터화된 객체를 반환한다.
- `CountVectorizer`의 입력 파라미터
  - `max_df` : 전체 문서 중 너무 높은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터. 정수 값일 경우, 그 개수 이하로 나타나는 단어만 피처로 추출하며, 0에서 1 사이의 소수값일 경우, 그 비율만큼의 단어만 피처로 추출한다.
  - `min_df` : 전체 문서 중 너무 낮은 빈도수를 가지는 단어 피처를 제외하기 위한 파라미터
  - `max_features` : 추출하는 피처의 개수를 제한한다. 정수값을 입력한다.
  - `stop_words` : 'english'로 지정하면 영어의 스톱 워드로 지정된 단어를 추출에서 제외한다.
  - `n_gram_range` : BOW 모델의 순서를 어느정도 보강하기 위한 n_gram의 범위를 튜플 형태로 입력한다.
  - `analyzer` : 피처 추출을 수행할 단위를 지정하는 파라미터로 디폴트는 'word'이다.
  - `token_patter` : 토큰화를 수행하는 정규 표현식 패턴을 지정한다. 디폴트 값은 '\b\w\w+\b'로 공백 또는 개행 문자 등으로 구분된 단어를 토큰으로 분리한다.
  - `tokenizer` : 토큰화를 별도의 커스텀 함수로 진행할 경우 적용한다.
- `CountVectorizer`의 텍스트 피처 벡터화 방법
  1. **사전 데이터 가공** : 영어의 경우 모든 문자를 소문자로 변경하는 등 전처리 작업을 수행한다.
  2. **토큰화** : 각 단어를 토큰화한다.
  3. **텍스트 정규화** : Stop Words 필터링을 수행한다.
  4. **피처 벡터화** : 입력된 파라미터를 바탕으로 토큰화된 단어를 피처로 추출하고 단어 빈도수 벡터 값을 적용한다.
- 사이킷런에서 TF-IDF 벡터화는 `TfidVectorizer` 클래스를 이용하며, 파라미터와 변환 방법은 `CountVectorizer`와 동일하다.

### | **BOW 벡터화를 위한 희소 행렬**
- 사이킷런의 `CountVectorizer`/`TfidVectorizer`를 이용해 텍스트를 피처 단위로 벡터화하면 CSR 형태의 희소 행렬을 반환한다.
- 대규모 행렬의 대부분의 값을 0이 차지하는 행렬을 희소 행렬이라고 하는데, BOW 형태를 가진 언어 모델의 피처 벡터화 결과는 대부분 희소 행렬이다.
- 희소 행렬 형태의 데이터셋은 불필요한 0 값이 메모리 공간에 할당되어 메모리 공간이 많이 필요하며, 행렬의 크기가 커서 연산 시에 시간이 많이 소모된다.
- 따라서 희소 행렬을 물리적으로 적은 메모리 공간을 차지하도록 변환해야 한다. 대표적인 방법으로 COO 형식과 CSR 형식이 있다.

### | **희소 행렬 - COO 형식**
- **COO(Coordinate) 형식**은 0이 아닌 데이터만 별도의 데이터 배열에 저장하고, 그 위치를 별도의 배열에 저장하는 방식이다.
- 사이파이(Scipy)의 `sparse` 패키지는 희소 행렬 변환을 위한 다양한 모듈을 제공한다. 이를 이용해 COO 형식으로 희소 행렬 변환을 수행해보자.
- 다음 예제는 밀집 행렬을 만든 후, COO 형식으로 희소행렬로 변환하는 문제이다.

In [24]:
import numpy as np
from scipy import sparse

In [26]:
# 밀집 행렬 생성
dense = np.array([[ 3, 0, 1 ], [0, 2, 0 ]])

In [27]:
data = np.array([3,1,2]) # 0이 아닌 데이터 추출

# 행 위치와 열 위치를 각각 array로 생성 
row_pos = np.array([0,0,1])
col_pos = np.array([0,2,1])

# sparse 패키지의 coo_matrix를 이용하여 COO 형식으로 희소 행렬 생성
sparse_coo = sparse.coo_matrix((data, (row_pos,col_pos)))

In [28]:
# 다시 밀집 형태의 행렬로 출력
sparse_coo.toarray()

array([[3, 0, 1],
       [0, 2, 0]])

### | **희소 행렬 - CSR 형식**
- CSR(Compressed Sparse Row) 형식은 COO 형식이 행과 열의 위치를 나타내기 위해 반복적인 위치 데이터를 사용해야 하는 문제점을 해결한 방법이다.
- 행 위치 배열 내에 있는 고유한 값의 시작 위치만 다시 별도의 위치 배열로 가지는 변환 방식을 의미한다.
- 고유 값의 시작 위치만 알고 있다면 얼마든지 행 위치 배열을 다시 만들 수 있어 COO 방식보다 메모리가 적게 들고 빠른 영상이 가능하다.

In [30]:
from scipy import sparse

dense2 = np.array([[0,0,1,0,0,5],
             [1,4,0,3,2,5],
             [0,6,0,3,0,0],
             [2,0,0,0,0,0],
             [0,0,0,7,0,8],
             [1,0,0,0,0,0]])

In [31]:
# 0 이 아닌 데이터 추출
data2 = np.array([1, 5, 1, 4, 3, 2, 5, 6, 3, 2, 7, 8, 1])

# 행 위치와 열 위치를 각각 array로 생성 
row_pos = np.array([0, 0, 1, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5])
col_pos = np.array([2, 5, 0, 1, 3, 4, 5, 1, 3, 0, 3, 5, 0])

# COO 형식으로 변환 
sparse_coo = sparse.coo_matrix((data2, (row_pos,col_pos)))

In [32]:
# CRS
row_pos_ind = np.array([0, 2, 7, 9, 10, 12, 13]) # 행 위치 배열의 고유한 값들의 시작 위치 인덱스를 배열로 생성

sparse_csr = sparse.csr_matrix((data2, col_pos, row_pos_ind))

In [33]:
print('COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_coo.toarray())

COO 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


In [35]:
print('CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인')
print(sparse_csr.toarray())

CSR 변환된 데이터가 제대로 되었는지 다시 Dense로 출력 확인
[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]
